compiler/rustc_typeck/src/check/callee.rs

```diff
@@ -390,17 +390,23 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 (fn_sig, Some(def_id))
             }
             ty::FnPtr(sig) => (sig, None),
-            ref t => {
+            _ => {
                 let mut unit_variant = None;
-                let mut removal_span = call_expr.span;
-                if let ty::Adt(adt_def, ..) = t
-                    && adt_def.is_enum()
-                    && let hir::ExprKind::Call(expr, _) = call_expr.kind
+                if let hir::ExprKind::Path(qpath) = &callee_expr.kind
+                    && let Res::Def(def::DefKind::Ctor(kind, def::CtorKind::Const), _)
+                        = self.typeck_results.borrow().qpath_res(qpath, callee_expr.hir_id)
+                    // Only suggest removing parens if there are no arguments
+                    && arg_exprs.is_empty()
+                    && let Ok(path) = self.tcx.sess.source_map().span_to_snippet(callee_expr.span)
```

Comment:
I would like to use `tcx.def_path_str` instead of `span_to_snippet`, but that does not preserve aliases, and causes regressions in `src/test/ui/type-alias-enum-variants/incorrect-variant-form-through-alias-caught.rs` by changing `Alias::Unit` to `Enum::Unit` :cry:

Reply:
I could possibly find a way to format `qpath` into a string by joining all of its segments, but that's a bit overkill and this PR does not represent a regression in the status quo.

Reply:
It wouldn't work well with improperly formatted code, right? If the path is a macro expansion what would happen..? To me I don't think they should preserve the alias path.

Reply:
Hm, you're right. If we had like

```
path::
split::
up::
over_many_lines
```

then we shouldn't use that in diagnostics.

Reply:
I just used `rustc_hir_pretty` which does that for us. Avoids weirdness like:

```rust
enum Enum { Braced {} }
type Alias = Enum;

fn main() {
    Alias::
    Braced;
}
```

```
error[[E0533]](https://doc.rust-lang.org/stable/error-index.html#E0533): expected unit struct, unit variant or constant, found struct variant `Alias::
                  Braced`
 --> src/main.rs:5:5
  |
5 | /     Alias::
6 | |     Braced;
  | |__________^
```

Reply:
Nice, please add that as a test.

Reply:
Done